### PR TITLE
[Heartbeat] Fix scheduler job type limit algorithm

### PR DESCRIPTION
**Pull request by andrewvc**
## What does this PR do?

Previously heartbeat would break when running with `mode: all` since that would create multiple terminal jobs. These would all attempt to release from the limit semaphore, when only the last one should.

This also refactors recursive job running into an OO type structure to make things more readable.

## Why is it important?

Causes a panic:

```
panic: semaphore: released more than held

goroutine 100 [running]:
golang.org/x/sync/semaphore.(*Weighted).Release(0xc00011c0f0, 0x1)
	/Users/dominiqueclarke/go/pkg/mod/golang.org/x/sync@v0.0.0-20200317015054-43a5402ce75a/semaphore/semaphore.go:103 +0xed
github.com/elastic/beats/v7/heartbeat/scheduler.(*Scheduler).runRecursiveTask(0xc000720000, 0x1101964a0, 0xc00073c440, 0xc00000ec20, 0xc000058a20, 0xc00011c0f0, 0xc04114cb9b52d7b0, 0x346f4dc, 0x1110a1b60)
	/Users/dominiqueclarke/dev/beats/heartbeat/scheduler/scheduler.go:303 +0x2e5
created by github.com/elastic/beats/v7/heartbeat/scheduler.(*Scheduler).runRecursiveTask
	/Users/dominiqueclarke/dev/beats/heartbeat/scheduler/scheduler.go:300 +0x246
```
With this config

```yaml
output.elasticsearch:
  hosts: ["localhost:9200"]
  username: "elastic"
  password: "[redacted]"
heartbeat.jobs.http.limit: 1
heartbeat.monitors:
- type: http
  id: httpcheck
  name: HTTP_CHECK
  urls: 'https://news.google.com'
  schedule: '@every 1m'
  mode: all
- type: http
  id: localhost
  name: localhost
  urls: 'http://localhost:8080'
  schedule: '@every 1m'
  mode: all
```
## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
~- [ ] I have made corresponding changes to the documentation~
~- [ ] I have made corresponding change to the default configuration files~
- [x] I have added tests that prove my fix is effective or that my feature works
~- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.~

## How to test this PR locally

Use the config above

### CODE REVIEW

File: heartbeat/scheduler/scheduler.go [hunk 1]
```diff
@@ -242,65 +243,94 @@ func (s *Scheduler) runOnce(runAt time.Time, taskFn timerqueue.TimerTaskFn) {
...
 			// We can discard the started at times for continuations as those are
 			// irrelevant
-			go s.runRecursiveTask(jobCtx, cont, wg, jobSem)
+			go sj.runTask(cont)
 		}
-		if jobSem != nil && len(continuations) == 0 {
-			jobSem.Release(1)
+		// There is always at least 1 task (the current one), if that's all, then we know
+		// there are no other jobs active or pending, and we can release the jobLimitSem
+		if sj.jobLimitSem != nil && sj.activeTasks.Load() == 1 {
```
Comment (andrewvc) [inline heartbeat/scheduler/scheduler.go:332]: This is the core of the fix, being able to count the actual number of active tasks

File: heartbeat/scheduler/scheduler_test.go [hunk 2]
```diff
@@ -311,7 +337,8 @@ func TestScheduler_runRecursiveJob(t *testing.T) {
 					taskArr = append(taskArr, num)
 				})
 				go func(tff TaskFunc) {
-					s.runRecursiveJob(context.Background(), tff, jobType)
+					sj := newSchedJob(context.Background(), s, "atype", tff)
```
Comment (vigneshshanmugam) [inline heartbeat/scheduler/scheduler_test.go:340]: Need to pass correct jobType here for the limit semaphore to apply.

File: x-pack/heartbeat/monitors.d/plaintodos.yml [hunk 3]
```diff
@@ -0,0 +1,11 @@
+- name: Todos
```
Comment (vigneshshanmugam) [inline x-pack/heartbeat/monitors.d/plaintodos.yml:1]: added by mistake?

File: x-pack/heartbeat/monitors.d/todos.yml [hunk 4]
```diff
@@ -0,0 +1,35 @@
+- id: 2d95e02d-0e6e-4654-b04e-7e38d273c5f3
```
Comment (vigneshshanmugam) [inline x-pack/heartbeat/monitors.d/todos.yml:1]: same

**Review by vigneshshanmugam — APPROVED** [review]
LGTM and looks much cleaner than before.

**Review by vigneshshanmugam — CHANGES_REQUESTED** [review]
Did a test locally, turned out to be not working as expected. Its not getting limited even after setting limit to 1

File: heartbeat/scheduler/scheduler_test.go (hunk 2 above)
Comment (andrewvc) [reply heartbeat/scheduler/scheduler_test.go:340]: 🤦

**Comment by andrewvc** [comment]
@vigneshshanmugam I believe I've addressed all PR feedback now

**Comment by andrewvc** [comment]
I take that back, I didn't see your final comment, will try a local test.

**Comment by andrewvc** [comment]
@vigneshshanmugam I see it working with the following config. I've pushed up a commit with WARN logging showing the locking. You can see that with a limit of one only one monitor can have acquired the lock

```
2021-08-23T19:34:32.996-0500    WARN    scheduler/scheduler.go:209      Run my-monitor | &{1 0 {0 0} {{0 0 0 <nil>} 0}} limit
2021-08-23T19:34:32.996-0500    WARN    scheduler/schedjob.go:46        TRY-BLOCKING-ACQUIRE ? my-monitor
2021-08-23T19:34:32.996-0500    WARN    scheduler/schedjob.go:48        ACQUIRED + my-monitor
2021-08-23T19:34:32.996-0500    WARN    scheduler/scheduler.go:209      Run alt-monitor | &{1 1 {0 0} {{0 0 0 <nil>} 0}} limit
2021-08-23T19:34:32.996-0500    WARN    scheduler/schedjob.go:46        TRY-BLOCKING-ACQUIRE ? alt-monitor
2021-08-23T19:34:32.996-0500    INFO    cfgfile/reload.go:164   Config reloader started
2021-08-23T19:34:32.997-0500    INFO    cfgfile/reload.go:224   Loading of config files completed.
2021-08-23T19:34:33.164-0500    WARN    scheduler/schedjob.go:105       RELEASED + my-monitor
2021-08-23T19:34:33.164-0500    WARN    scheduler/scheduler.go:213      End my-monitor
2021-08-23T19:34:33.164-0500    WARN    scheduler/schedjob.go:48        ACQUIRED + alt-monitor
2021-08-23T19:34:33.189-0500    WARN    scheduler/schedjob.go:105       RELEASED + alt-monitor
2021-08-23T19:34:33.189-0500    WARN    scheduler/scheduler.go:213      End alt-monitor
```


sample config used is:

```yaml
heartbeat.config.monitors:
  path: ${path.config}/monitors.d/*.yml
  reload.enabled: false
  reload.period: 5s
heartbeat.jobs.http.limit: 1
heartbeat.monitors:
- type: http
  id: my-monitor
  name: My Monitor
  urls: ["http://www.google.com"]
  mode: all
  schedule: '@every 10s'
- type: http
  id: alt-monitor
  mode: all
  name: Alt Monitor
  urls: ["http://www.elastic.co"]
  schedule: '@every 10s'

setup.template.settings:
  index.number_of_shards: 1
  index.codec: best_compression
setup.kibana:
output.console: ~
processors:
  - add_observer_metadata:
```

**Comment by vigneshshanmugam** [comment]
@andrewvc Sorry for that, my bad I had my acquire print statements in a wrong place before the acquire which made me think it was acquiring way too sooner, But I did test now properly and can see it working fine.

**Review by vigneshshanmugam — APPROVED** [review]
LGTM. Need to remove this file though - `x-pack/heartbeat/out`

**Comment by andrewvc** [comment]
@Mergifyio backport 7.x

**Comment by andrewvc** [comment]
@Mergifyio backport 7.15

**Pull request merged**